### ISSUE TITLE
Tutorial or docs on asynchronous streaming interface (w/ Express)

### ISSUE THREAD

**Issue by jeffersonlam** [linkedin/dustjs#567]
### Use Case
I'm working on an app using Node/Express and Dust. We're making a bunch of API calls in the server using Express, and then rendering the view with the data at the end. In doing so, however long it takes for the API calls to complete, that's how long it takes for the page to load. This isn't ideal, of course. Instead, we want to asynchronously pipe the data into the view from the server as it's loaded. 

### The Problem
I'm trying to figure out how to use the Dust streaming interface. I haven't found much information on how to implement the Dust streaming interface, specifically with a Node/Express framework. All of the docs I've seen only provide small code snippets of how bits and pieces work, but none provide the entire context.

Could you guys create docs, or code samples, of how to implement this? Is there a 'Hello World' of the Dust streaming interface?

**Comment by sethkinast** [comment]
I can work on a simpler example, but you might look at Adaro, part of the Kraken framework-- they have streaming support.

The code is pretty helpful (goofy names for streams notwithstanding): https://github.com/krakenjs/adaro/blob/v2.x/lib/engine.js#L138

The short version, while I work on a concrete example, is that you should pipe the stream to `res` in your Express handler:

```
app.get("/", function(req, res) {
  dust.stream("template", {"context": "something"}).pipe(res);
});
```

**Comment by sethkinast** [comment]
Oh, and the other caveat is that your async loaders need to be implemented as Dust context helpers so that Dust can know that there is asynchronous data that will be coming in.

**Comment by sethkinast** [comment]
Also Hoffman:

https://github.com/diffsky/hoffman/

**Comment by rragan** [comment]
Would page on dustjs.com devoted to external links related to dust be a good resource?

**Comment by jeffersonlam** [comment]
I didn't even realize that www.dustjs.com was a website. It doesn't seem to pop up when you search for DustJS on Google. Yes, a list of resources to help get started with Dust would be very, very helpful. There doesn't seem to be one single place that I can rely on for Dust documentation and resources; I have to google DustJS and keep my fingers crossed. 

It actually looks like www.dustjs.com and this link http://linkedin.github.io/dustjs/ have the same content. The latter doesn't have any CSS though, and the links are broken. Is dustjs.com the "real" version of this content? The latter is currently linked to from the readme. 

I've heard that Express doesn't natively support Dust's streaming interface; do the 3rd party modules Hoffman and Adaro help bridge that gap? Is a module like that required to support streaming in Express, or can I do without them?

**Comment by sethkinast** [comment]
Express does not natively support any sort of streaming, that's correct. The Express model is based around `res.render`, which passes a fully-rendered template to the response.

Hoffman does actively bridge the gap by providing `res.stream` that works just like `res.render` and pipes directly to `res` if you want, or gives you the stream back so you can manipulate it before piping.

Adaro takes a config option to turn on streaming, but makes you use `res.render` and gives you a `stream` object in the callback that you then have to manually pipe to `res` yourself.

Both engines set themselves up to handle your views directory and load Dust partials correctly.

**Comment by sethkinast** [comment]
#575 has been pulled in and shows some simple streaming examples. How do those look?

**Comment by jeffersonlam** [comment]
These look great! This type of simple "Hello World" example is exactly what I was looking for. Now I can see how a DustJS project can be set up. 

Some comments, as a newcomer to DustJS:

#### basic-node
This is great. By looking at these 12 lines of code in I can gain a basic understanding of a simple app built with Node and Dust, and the concepts behind it (eg `dust.compile`, `dust.loadSource`, `dust.render`)

Perhaps this will become more clear as you guys roll out more docs (I noticed there are more issues on docs that just popped up), but what's the purpose of compiling a template? In my current project we aren't doing `dust.compile` or `dust.loadSource`, we're just using Express's `res.render` function and Dust as the view engine. 

#### basic-browser
I'm running the `bower install` command in the readme, but it's giving me this error:
![screen shot 2015-03-27 at 11 47 56 am](https://cloud.githubusercontent.com/assets/6686963/6900556/e2f0f0d4-d6cc-11e4-9701-da7a2fdfed83.png)

#### streaming
This is cool - this is my first time seeing streaming in action. I didn't know you could literally pipe in a whole website. 

I experimented a bit and added some text above and below the partial tag, like so:
```
<h1>hello</h1>
  {>"dust-site"/}
<h1>end</h1>
```
I noticed that the text above the partial gets rendered, but the text below it doesn't. Is that expected behavior? Does Dust load pages linearly with streaming, or is it possible to asynchronously populate portions of the page, no matter where they are?

#### streaming-hoffman
I'm a little unclear on what exactly Hoffman is and what it does, but putting this example and the `streaming` example side by side, it seems to simplify the syntax a bit. Is Hoffman a preferred way to handle Dust streaming?

#### amd
I don't have a use case for this, as I'm not using Require.js. Is it common to pair Require.JS and Dust together?

**Comment by sethkinast** [comment]
Thanks so much for taking the time to look at all the examples. This will help me greatly in refining them.

`basic-node` - Whatever `view engine` you have hooked up to Express is doing the compiling for you. Compiling is a slow step (100x slower than rendering) so you want to do it once. Dust stores the compiled output for you so you can rerender the same template again without having to recompile.

`basic-browser` - You're right; it assumes that you installed dust from NPM before running the examples, but you've just downloaded the example folder. Will fix.

`streaming` - Both tags will be there, but the dust site includes a </body> tag and it's probably just not rendering anything after that. Maybe I should stream something simpler. Will enhance.

`streaming-hoffman` - Hoffman is a view engine for express; it simplifies the syntax and manages the compilation for you. `adaro`, `hoffman`, `klei-dust`, and `consolidate` are all view engines for Dust that you can use that support different functionality.

`amd` - useful if you were building a single-page app, or something that did a lot of dynamic rerendering. We just added AMD / require.js support as of Dust 2.6.0.

**Comment by jeffersonlam** [comment]
Of course, glad to be of help. And thank you for taking the time to make these examples.

I have been more descriptive with my comments on `streaming`, sorry. What I meant to say was that during the timeout, I can see the text above the partial, but not below it. 
![screen shot 2015-03-30 at 2 49 34 pm](https://cloud.githubusercontent.com/assets/6686963/6905216/19abd37c-d6ec-11e4-9780-25618907d649.png)
After the partial gets loaded, I can then see the text below (and above) the partial. 
![screen shot 2015-03-30 at 2 49 37 pm](https://cloud.githubusercontent.com/assets/6686963/6905218/1cf969f4-d6ec-11e4-8efb-85ac54337745.png)
![screen shot 2015-03-30 at 2 49 39 pm](https://cloud.githubusercontent.com/assets/6686963/6905219/1e5e5d04-d6ec-11e4-94ab-11d7b6ab467c.png)
Is that expected behavior from Dust, or should the text below the partial be visible even when the partial hasn't been streamed in yet? I've also tried replacing that URL to a simple API request (Open Weather Map API) to return a JSON object, and the behavior is the same.

**Comment by sethkinast** [comment]
There are two levels of streaming in play here. The first is at the Dust level. The second is at the browser level. When Dust encounters something asynchronous, it can continue rendering the template without waiting for that call to complete. So say you had something that was basically:

    Hello
    ...async call 1, takes 1 second
    ...async call 2, takes 2 seconds
    ...async call 3, takes 3 seconds
    End

Dust would finish rendering this template in 3 seconds because all calls would be made in parallel.

If you used `dust.render`, at that time the completed template would be delivered to the callback and sent to the browser / smashed into some innerHTML, depending on how you were running Dust.

If you use `dust.stream`, Dust sends `Hello`, waits 1 second, sends the first async block, waits 1 second, sends the next one, etc. But you must send HTML in order. Dust can't send the HTML of the page out-of-order, because browsers can't assemble HTML out-of-order. However, this still has the advantage of letting you stream your page and get the browser downloading Javascript / CSS, while you wait on your async calls (to an authentication server or something) to complete, and it doesn't force your page to be as slow as the slowest call (especially if the slow call is down low on the page, you could stream out everything on top first).

If you had a complete page with something in, say, the sidebar that loaded slowly, you would want to wire that up on the client side. So after your page has loaded, you could have something like:

    <script>
    dust.render("mySidebarWidget", { data: $.getJSON('/myAPI/') }, function(err, out) {
      $("#theWidget").html(out);
    });
    </script>

Dust can accept a Deferred object from a jQuery AJAX call directly-- this template will be rendered with the data from that call once the AJAX call completes.

**Comment by sethkinast** [comment]
I do think Dust streaming is underutilized and a big part of that is documentation, so we'll definitely work on fleshing that out.

**Comment by sethkinast** [comment]
I've added another example of streaming to show what's going on behind the scenes. Thanks for all the feedback!
